File: backend/app/models/recipe.py
"""
Recipe Model
Comprehensive recipe management with ingredients, instructions, nutrition, and social features
"""

from sqlalchemy import (
    Column, Integer, String, Boolean, DateTime, Text, 
    JSON, Float, Enum as SQLEnum, ForeignKey, Table
)
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from sqlalchemy.dialects.postgresql import UUID, ARRAY
import uuid
import enum
from datetime import datetime
from typing import Dict, List, Optional, Any

from app.core.database import Base


class MealType(str, enum.Enum):
    """Meal type enumeration"""
    BREAKFAST = "breakfast"
    BRUNCH = "brunch"
    LUNCH = "lunch"
    DINNER = "dinner"
    SNACK = "snack"
    DESSERT = "dessert"
    APPETIZER = "appetizer"
    BEVERAGE = "beverage"


class DifficultyLevel(str, enum.Enum):
    """Recipe difficulty level"""
    VERY_EASY = "very_easy"
    EASY = "easy"
    MEDIUM = "medium"
    HARD = "hard"
    VERY_HARD = "very_hard"


class RecipeStatus(str, enum.Enum):
    """Recipe status enumeration"""
    DRAFT = "draft"
    PUBLISHED = "published"
    PRIVATE = "private"
    ARCHIVED = "archived"
    FLAGGED = "flagged"


# Association table for recipe tags
recipe_tags = Table(
    'recipe_tags',
    Base.metadata,
    Column('recipe_id', UUID(as_uuid=True), ForeignKey('recipes.id'), primary_key=True),
    Column('tag_id', UUID(as_uuid=True), ForeignKey('tags.id'), primary_key=True)
)


class Recipe(Base):
    """
    Recipe model with comprehensive recipe information
    """
    __tablename__ = "recipes"

    # Primary identification
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4, index=True)
    title = Column(String(200), nullable=False, index=True)
    slug = Column(String(250), unique=True, nullable=False, index=True)
    description = Column(Text, nullable=True)
    
    # Author and visibility
    author_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    status = Column(SQLEnum(RecipeStatus), default=RecipeStatus.PUBLISHED, nullable=False)
    visibility = Column(String(20), default="public")  # public, private, friends
    
    # Recipe classification
    meal_type = Column(SQLEnum(MealType), nullable=True)
    cuisine_type = Column(String(50), nullable=True)  # italian, mexican, asian, etc.
    difficulty_level = Column(SQLEnum(DifficultyLevel), default=DifficultyLevel.MEDIUM)
    
    # Timing and servings
    prep_time = Column(Integer, nullable=True)  # minutes
    cook_time = Column(Integer, nullable=True)  # minutes
    total_time = Column(Integer, nullable=True)  # minutes
    servings = Column(Integer, default=4, nullable=False)
    
    # Instructions and ingredients
    instructions = Column(JSON, nullable=False)  # List of step objects
    ingredients_list = Column(JSON, nullable=False)  # List of ingredient objects
    equipment_needed = Column(ARRAY(String), default=list)
    
    # Media
    main_image_url = Column(String(500), nullable=True)
    images = Column(JSON, default=list)  # List of image URLs
    video_url = Column(String(500), nullable=True)
    
    # Nutrition information
    calories_per_serving = Column(Integer, nullable=True)
    protein_per_serving = Column(Float, nullable=True)
    carbs_per_serving = Column(Float, nullable=True)
    fat_per_serving = Column(Float, nullable=True)
    fiber_per_serving = Column(Float, nullable=True)
    sugar_per_serving = Column(Float, nullable=True)
    sodium_per_serving = Column(Float, nullable=True)
    
    # Cost and budget information
    estimated_cost = Column(Float, nullable=True)
    cost_per_serving = Column(Float, nullable=True)
    currency = Column(String(3), default="USD")
    is_budget_friendly = Column(Boolean, default=False)
    
    # Dietary information
    is_vegetarian = Column(Boolean, default=False)
    is_vegan = Column(Boolean, default=False)
    is_gluten_free = Column(Boolean, default=False)
    is_dairy_free = Column(Boolean, default=False)
    is_nut_free = Column(Boolean, default=False)
    is_low_carb = Column(Boolean, default=False)
    is_keto = Column(Boolean, default=False)
    is_paleo = Column(Boolean, default=False)
    is_halal = Column(Boolean, default=False)
    is_kosher = Column(Boolean, default=False)
    
    # Recipe source and generation
    source_type = Column(String(20), default="ai_generated")  # ai_generated, user_created, imported
    source_url = Column(String(500), nullable=True)
    ai_model_used = Column(String(50), nullable=True)
    generation_prompt = Column(Text, nullable=True)
    
    # Social and engagement
    likes_count = Column(Integer, default=0)
    views_count = Column(Integer, default=0)
    shares_count = Column(Integer, default=0)
    comments_count = Column(Integer, default=0)
    favorites_count = Column(Integer, default=0)
    cooked_count = Column(Integer, default=0)  # How many times it was cooked
    
    # Recipe rating
    average_rating = Column(Float, default=0.0)
    ratings_count = Column(Integer, default=0)
    
    # SEO and searchability
    meta_title = Column(String(255), nullable=True)
    meta_description = Column(Text, nullable=True)
    keywords = Column(ARRAY(String), default=list)
    
    # Recipe variants and alternatives
    parent_recipe_id = Column(UUID(as_uuid=True), ForeignKey("recipes.id"), nullable=True)
    variant_type = Column(String(50), nullable=True)  # low_calorie, gluten_free, etc.
    
    # Seasonal and special occasions
    seasonal_tags = Column(ARRAY(String), default=list)  # spring, summer, fall, winter
    occasion_tags = Column(ARRAY(String), default=list)  # birthday, christmas, etc.
    
    # Recipe feedback and improvements
    feedback_summary = Column(JSON, default=dict)
    improvement_suggestions = Column(JSON, default=list)
    
    # Analytics and tracking
    performance_score = Column(Float, default=0.0)  # Overall recipe performance
    engagement_rate = Column(Float, default=0.0)
    conversion_rate = Column(Float, default=0.0)  # Views to cooked ratio
    
    # Timestamps
    created_at = Column(DateTime, default=func.now(), nullable=False)
    updated_at = Column(DateTime, default=func.now(), onupdate=func.now(), nullable=False)
    published_at = Column(DateTime, nullable=True)
    last_cooked_at = Column(DateTime, nullable=True)
    
    # Relationships
    author = relationship("User", back_populates="recipes")
    ingredients = relationship("RecipeIngredient", back_populates="recipe", cascade="all, delete-orphan")
    reviews = relationship("Review", back_populates="recipe", cascade="all, delete-orphan")
    favorites = relationship("Favorite", back_populates="recipe", cascade="all, delete-orphan")
    variations = relationship("Recipe", remote_side=[id])
    tags = relationship("Tag", secondary=recipe_tags, back_populates="recipes")
    nutrition_analysis = relationship("NutritionAnalysis", back_populates="recipe", uselist=False)
    
    def __repr__(self) -> str:
        return f"<Recipe(id={self.id}, title={self.title}, author_id={self.author_id})>"
    
    @property
    def total_time_minutes(self) -> int:
        """Calculate total time from prep and cook time"""
        if self.total_time:
            return self.total_time
        prep = self.prep_time or 0
        cook = self.cook_time or 0
        return prep + cook
    
    @property
    def difficulty_score(self) -> int:
        """Get numeric difficulty score"""
        difficulty_scores = {
            DifficultyLevel.VERY_EASY: 1,
            DifficultyLevel.EASY: 2,
            DifficultyLevel.MEDIUM: 3,
            DifficultyLevel.HARD: 4,
            DifficultyLevel.VERY_HARD: 5
        }
        return difficulty_scores.get(self.difficulty_level, 3)
    
    def calculate_nutrition_per_serving(self) -> Dict[str, float]:
        """Calculate nutrition values per serving"""
        return {
            "calories": self.calories_per_serving or 0,
            "protein": self.protein_per_serving or 0,
            "carbs": self.carbs_per_serving or 0,
            "fat": self.fat_per_serving or 0,
            "fiber": self.fiber_per_serving or 0,
            "sugar": self.sugar_per_serving or 0,
            "sodium": self.sodium_per_serving or 0,
        }
    
    def get_dietary_flags(self) -> List[str]:
        """Get list of dietary restriction flags"""
        flags = []
        if self.is_vegetarian: flags.append("vegetarian")
        if self.is_vegan: flags.append("vegan")
        if self.is_gluten_free: flags.append("gluten_free")
        if self.is_dairy_free: flags.append("dairy_free")
        if self.is_nut_free: flags.append("nut_free")
        if self.is_low_carb: flags.append("low_carb")
        if self.is_keto: flags.append("keto")
        if self.is_paleo: flags.append("paleo")
        if self.is_halal: flags.append("halal")
        if self.is_kosher: flags.append("kosher")
        return flags
    
    def increment_view_count(self) -> None:
        """Increment view count"""
        self.views_count += 1
    
    def increment_like_count(self) -> None:
        """Increment like count"""
        self.likes_count += 1
    
    def decrement_like_count(self) -> None:
        """Decrement like count"""
        if self.likes_count > 0:
            self.likes_count -= 1
    
    def increment_cooked_count(self) -> None:
        """Increment cooked count"""
        self.cooked_count += 1
        self.last_cooked_at = datetime.utcnow()
    
    def update_rating(self, new_rating: float) -> None:
        """Update average rating with new rating"""
        if self.ratings_count == 0:
            self.average_rating = new_rating
            self.ratings_count = 1
        else:
            total_rating = self.average_rating * self.ratings_count
            self.ratings_count += 1
            self.average_rating = (total_rating + new_rating) / self.ratings_count
    
    def calculate_performance_score(self) -> float:
        """Calculate overall recipe performance score"""
        # Weighted scoring based on various metrics
        weights = {
            "rating": 0.3,
            "engagement": 0.25,
            "conversion": 0.2,
            "popularity": 0.15,
            "freshness": 0.1
        }
        
        # Rating score (0-1)
        rating_score = (self.average_rating / 5.0) if self.ratings_count > 0 else 0
        
        # Engagement score (likes + favorites per view)
        engagement_score = 0
        if self.views_count > 0:
            engagement_score = min((self.likes_count + self.favorites_count) / self.views_count, 1.0)
        
        # Conversion score (cooked per view)
        conversion_score = 0
        if self.views_count > 0:
            conversion_score = min(self.cooked_count / self.views_count, 1.0)
        
        # Popularity score (based on total interactions)
        total_interactions = self.likes_count + self.favorites_count + self.cooked_count
        popularity_score = min(total_interactions / 100.0, 1.0)  # Normalize to 100 interactions
        
        # Freshness score (newer recipes get higher scores)
        days_since_creation = (datetime.utcnow() - self.created_at).days
        freshness_score = max(1.0 - (days_since_creation / 365.0), 0)  # Decay over a year
        
        # Calculate weighted score
        score = (
            weights["rating"] * rating_score +
            weights["engagement"] * engagement_score +
            weights["conversion"] * conversion_score +
            weights["popularity"] * popularity_score +
            weights["freshness"] * freshness_score
        )
        
        self.performance_score = score
        return score


class RecipeIngredient(Base):
    """
    Association table for recipe ingredients with quantities
    """
    __tablename__ = "recipe_ingredients"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    recipe_id = Column(UUID(as_uuid=True), ForeignKey("recipes.id"), nullable=False)
    ingredient_id = Column(UUID(as_uuid=True), ForeignKey("ingredients.id"), nullable=False)
    
    # Quantity and measurement
    quantity = Column(Float, nullable=False)
    unit = Column(String(50), nullable=False)  # cup, tbsp, oz, g, etc.
    preparation = Column(String(100), nullable=True)  # diced, minced, etc.
    
    # Additional properties
    is_optional = Column(Boolean, default=False)
    substitutes = Column(ARRAY(String), default=list)
    notes = Column(Text, nullable=True)
    
    # Order in the recipe
    order_index = Column(Integer, default=0)
    
    # Timestamps
    created_at = Column(DateTime, default=func.now())
    
    # Relationships
    recipe = relationship("Recipe", back_populates="ingredients")
    ingredient = relationship("Ingredient", back_populates="recipe_ingredients")
    
    def __repr__(self) -> str:
        return f"<RecipeIngredient(recipe_id={self.recipe_id}, ingredient_id={self.ingredient_id})>"
    
    @property
    def display_text(self) -> str:
        """Get formatted ingredient text for display"""
        text = f"{self.quantity} {self.unit}"
        if self.ingredient:
            text += f" {self.ingredient.name}"
        if self.preparation:
            text += f", {self.preparation}"
        if self.is_optional:
            text += " (optional)"
        return text


class Tag(Base):
    """
    Tags for categorizing recipes
    """
    __tablename__ = "tags"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    name = Column(String(50), unique=True, nullable=False, index=True)
    description = Column(Text, nullable=True)
    category = Column(String(50), nullable=True)  # cuisine, dietary, occasion, etc.
    color = Column(String(7), nullable=True)  # Hex color code
    
    # Usage statistics
    usage_count = Column(Integer, default=0)
    
    # Timestamps
    created_at = Column(DateTime, default=func.now())
    
    # Relationships
    recipes = relationship("Recipe", secondary=recipe_tags, back_populates="tags")
    
    def __repr__(self) -> str:
        return f"<Tag(id={self.id}, name={self.name})>"
    
    def increment_usage(self) -> None:
        """Increment usage count"""
        self.usage_count += 1


class RecipeStep(Base):
    """
    Individual recipe steps for detailed instructions
    """
    __tablename__ = "recipe_steps"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    recipe_id = Column(UUID(as_uuid=True), ForeignKey("recipes.id"), nullable=False)
    
    # Step details
    step_number = Column(Integer, nullable=False)
    title = Column(String(200), nullable=True)
    instruction = Column(Text, nullable=False)
    
    # Timing and temperature
    duration = Column(Integer, nullable=True)  # minutes
    temperature = Column(Integer, nullable=True)  # degrees
    temperature_unit = Column(String(1), default="F")  # F or C
    
    # Media
    image_url = Column(String(500), nullable=True)
    video_url = Column(String(500), nullable=True)
    
    # Additional information
    tips = Column(Text, nullable=True)
    warnings = Column(Text, nullable=True)
    equipment = Column(ARRAY(String), default=list)
    
    # Timestamps
    created_at = Column(DateTime, default=func.now())
    
    # Relationships
    recipe = relationship("Recipe")
    
    def __repr__(self) -> str:
        return f"<RecipeStep(recipe_id={self.recipe_id}, step={self.step_number})>"

File: backend/models/Recipe.js
/**
 * Recipe Model
 * 
 * Represents a recipe in the AI Recipe Generator application.
 * Includes detailed recipe information, nutrition analysis,
 * cost breakdown, and user interactions.
 */

const mongoose = require('mongoose');

const ingredientSchema = new mongoose.Schema({
  name: {
    type: String,
    required: [true, 'Ingredient name is required'],
    trim: true
  },
  
  quantity: {
    type: Number,
    required: [true, 'Ingredient quantity is required'],
    min: [0, 'Quantity cannot be negative']
  },
  
  unit: {
    type: String,
    required: [true, 'Ingredient unit is required'],
    enum: [
      'cup', 'cups', 'tablespoon', 'tablespoons', 'teaspoon', 'teaspoons',
      'pound', 'pounds', 'ounce', 'ounces', 'gram', 'grams', 'kilogram', 'kilograms',
      'liter', 'liters', 'milliliter', 'milliliters', 'pint', 'pints', 'quart', 'quarts',
      'gallon', 'gallons', 'piece', 'pieces', 'slice', 'slices', 'clove', 'cloves',
      'bunch', 'bunches', 'package', 'packages', 'can', 'cans', 'bottle', 'bottles',
      'to taste', 'as needed', 'whole', 'half', 'quarter'
    ]
  },
  
  category: {
    type: String,
    enum: [
      'proteins', 'vegetables', 'fruits', 'grains', 'dairy', 'spices',
      'herbs', 'oils', 'condiments', 'nuts', 'legumes', 'beverages', 'other'
    ],
    default: 'other'
  },
  
  optional: {
    type: Boolean,
    default: false
  },
  
  substitutes: [{
    name: String,
    ratio: {
      type: String,
      default: '1:1'
    }
  }],
  
  // Cost information
  cost: {
    estimated: {
      type: Number,
      min: 0
    },
    
    currency: {
      type: String,
      default: 'USD'
    },
    
    pricePerUnit: Number,
    
    source: {
      type: String,
      enum: ['database', 'api', 'estimated', 'user'],
      default: 'estimated'
    }
  },
  
  // Nutrition per ingredient
  nutrition: {
    calories: Number,
    protein: Number, // grams
    carbs: Number,   // grams
    fat: Number,     // grams
    fiber: Number,   // grams
    sugar: Number,   // grams
    sodium: Number,  // milligrams
    cholesterol: Number, // milligrams
    vitamins: [{
      name: String,
      amount: Number,
      unit: String
    }],
    minerals: [{
      name: String,
      amount: Number,
      unit: String
    }]
  }
});

const instructionSchema = new mongoose.Schema({
  step: {
    type: Number,
    required: [true, 'Step number is required'],
    min: 1
  },
  
  instruction: {
    type: String,
    required: [true, 'Instruction text is required'],
    trim: true
  },
  
  duration: {
    type: Number, // in minutes
    min: 0
  },
  
  temperature: {
    value: Number,
    unit: {
      type: String,
      enum: ['F', 'C'],
      default: 'F'
    }
  },
  
  equipment: [String],
  
  tips: [String],
  
  images: [String], // URLs to step images
  
  warnings: [String] // Safety warnings or important notes
});

const nutritionSchema = new mongoose.Schema({
  perServing: {
    calories: {
      type: Number,
      min: 0
    },
    
    macros: {
      protein: { type: Number, min: 0 }, // grams
      carbs: { type: Number, min: 0 },   // grams
      fat: { type: Number, min: 0 }      // grams
    },
    
    micronutrients: {
      fiber: { type: Number, min: 0 },        // grams
      sugar: { type: Number, min: 0 },        // grams
      sodium: { type: Number, min: 0 },       // milligrams
      cholesterol: { type: Number, min: 0 },  // milligrams
      potassium: { type: Number, min: 0 },    // milligrams
      calcium: { type: Number, min: 0 },      // milligrams
      iron: { type: Number, min: 0 },         // milligrams
      vitaminA: { type: Number, min: 0 },     // IU
      vitaminC: { type: Number, min: 0 },     // milligrams
      vitaminD: { type: Number, min: 0 },     // IU
      vitaminE: { type: Number, min: 0 },     // milligrams
      vitaminK: { type: Number, min: 0 },     // micrograms
      vitaminB6: { type: Number, min: 0 },    // milligrams
      vitaminB12: { type: Number, min: 0 },   // micrograms
      folate: { type: Number, min: 0 },       // micrograms
      niacin: { type: Number, min: 0 },       // milligrams
      riboflavin: { type: Number, min: 0 },   // milligrams
      thiamin: { type: Number, min: 0 }       // milligrams
    }
  },
  
  total: {
    calories: Number,
    protein: Number,
    carbs: Number,
    fat: Number
  },
  
  healthScore: {
    type: Number,
    min: 0,
    max: 100
  },
  
  dietaryFlags: [{
    type: String,
    enum: [
      'vegetarian', 'vegan', 'gluten-free', 'dairy-free',
      'nut-free', 'egg-free', 'soy-free', 'shellfish-free',
      'kosher', 'halal', 'paleo', 'keto', 'low-carb',
      'low-fat', 'low-sodium', 'high-protein', 'high-fiber'
    ]
  }],
  
  allergens: [String],
  
  calculatedAt: {
    type: Date,
    default: Date.now
  },
  
  source: {
    type: String,
    enum: ['calculated', 'api', 'manual'],
    default: 'calculated'
  }
});

const costAnalysisSchema = new mongoose.Schema({
  total: {
    estimated: {
      type: Number,
      min: 0
    },
    
    currency: {
      type: String,
      default: 'USD'
    }
  },
  
  perServing: {
    type: Number,
    min: 0
  },
  
  breakdown: [{
    category: String,
    amount: Number,
    percentage: Number
  }],
  
  budgetFriendly: {
    type: Boolean,
    default: false
  },
  
  savingsOpportunities: [{
    suggestion: String,
    potentialSavings: Number
  }],
  
  calculatedAt: {
    type: Date,
    default: Date.now
  }
});

const recipeSchema = new mongoose.Schema({
  // Basic Information
  title: {
    type: String,
    required: [true, 'Recipe title is required'],
    trim: true,
    maxlength: [200, 'Title cannot exceed 200 characters']
  },
  
  description: {
    type: String,
    trim: true,
    maxlength: [1000, 'Description cannot exceed 1000 characters']
  },
  
  cuisine: {
    type: String,
    enum: [
      'american', 'italian', 'mexican', 'chinese', 'japanese',
      'indian', 'thai', 'french', 'mediterranean', 'middle-eastern',
      'korean', 'vietnamese', 'greek', 'spanish', 'british', 'fusion'
    ]
  },
  
  category: {
    type: String,
    enum: [
      'breakfast', 'lunch', 'dinner', 'snack', 'dessert',
      'appetizer', 'main-course', 'side-dish', 'soup', 'salad',
      'beverage', 'sauce', 'condiment', 'bread', 'pasta'
    ]
  },
  
  difficulty: {
    type: String,
    enum: ['easy', 'medium', 'hard'],
    default: 'medium'
  },
  
  // Timing Information
  timing: {
    prep: {
      type: Number, // minutes
      required: [true, 'Prep time is required'],
      min: [0, 'Prep time cannot be negative']
    },
    
    cook: {
      type: Number, // minutes
      required: [true, 'Cook time is required'],
      min: [0, 'Cook time cannot be negative']
    },
    
    total: {
      type: Number, // minutes
      min: [0, 'Total time cannot be negative']
    },
    
    inactive: {
      type: Number, // minutes (marinating, chilling, etc.)
      default: 0
    }
  },
  
  // Serving Information
  servings: {
    count: {
      type: Number,
      required: [true, 'Serving count is required'],
      min: [1, 'Must serve at least 1 person'],
      max: [50, 'Cannot serve more than 50 people']
    },
    
    size: String // e.g., "1 cup", "2 pieces"
  },
  
  // Recipe Content
  ingredients: {
    type: [ingredientSchema],
    required: [true, 'At least one ingredient is required'],
    validate: {
      validator: function(ingredients) {
        return ingredients && ingredients.length > 0;
      },
      message: 'Recipe must have at least one ingredient'
    }
  },
  
  instructions: {
    type: [instructionSchema],
    required: [true, 'At least one instruction is required'],
    validate: {
      validator: function(instructions) {
        return instructions && instructions.length > 0;
      },
      message: 'Recipe must have at least one instruction'
    }
  },
  
  // Equipment and Tools
  equipment: [{
    name: String,
    essential: {
      type: Boolean,
      default: true
    },
    alternatives: [String]
  }],
  
  // Media
  images: [{
    url: String,
    caption: String,
    isPrimary: {
      type: Boolean,
      default: false
    }
  }],
  
  video: {
    url: String,
    duration: Number, // seconds
    thumbnail: String
  },
  
  // Nutritional Information
  nutrition: nutritionSchema,
  
  // Cost Analysis
  costAnalysis: costAnalysisSchema,
  
  // Recipe Generation Information
  generatedFrom: {
    type: String,
    enum: ['image', 'ingredients-list', 'manual', 'ai-suggestion'],
    default: 'manual'
  },
  
  sourceIngredients: [String], // Original ingredients detected from image/list
  
  generationPrompt: String, // AI prompt used to generate recipe
  
  aiModel: String, // AI model used for generation
  
  // Creator and Attribution
  creator: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  },
  
  source: {
    type: String,
    enum: ['user-created', 'ai-generated', 'imported', 'community'],
    default: 'ai-generated'
  },
  
  originalSource: {
    url: String,
    name: String,
    attribution: String
  },
  
  // Social and Interaction Data
  ratings: [{
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    
    rating: {
      type: Number,
      min: 1,
      max: 5,
      required: true
    },
    
    review: String,
    
    helpful: [{
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    }],
    
    createdAt: {
      type: Date,
      default: Date.now
    }
  }],
  
  averageRating: {
    type: Number,
    min: 0,
    max: 5,
    default: 0
  },
  
  totalRatings: {
    type: Number,
    default: 0
  },
  
  // Usage Statistics
  stats: {
    views: {
      type: Number,
      default: 0
    },
    
    saves: {
      type: Number,
      default: 0
    },
    
    shares: {
      type: Number,
      default: 0
    },
    
    cooked: {
      type: Number,
      default: 0
    }
  },
  
  // Recipe Variations
  variations: [{
    name: String,
    description: String,
    modifications: [String],
    creator: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    }
  }],
  
  // Tags and Categories
  tags: [String],
  
  seasonal: [{
    type: String,
    enum: ['spring', 'summer', 'fall', 'winter']
  }],
  
  occasions: [{
    type: String,
    enum: [
      'everyday', 'weekend', 'holiday', 'party', 'date-night',
      'family-dinner', 'meal-prep', 'potluck', 'bbq', 'picnic'
    ]
  }],
  
  // Publication Status
  status: {
    type: String,
    enum: ['draft', 'published', 'private', 'archived'],
    default: 'published'
  },
  
  isPublic: {
    type: Boolean,
    default: true
  },
  
  featured: {
    type: Boolean,
    default: false
  },
  
  // Quality Assurance
  verified: {
    type: Boolean,
    default: false
  },
  
  qualityScore: {
    type: Number,
    min: 0,
    max: 100
  },
  
  flags: [{
    type: {
      type: String,
      enum: ['inappropriate', 'inaccurate', 'spam', 'copyright']
    },
    reason: String,
    reporter: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    createdAt: {
      type: Date,
      default: Date.now
    }
  }]
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Indexes for performance
recipeSchema.index({ title: 'text', description: 'text', tags: 'text' });
recipeSchema.index({ cuisine: 1, category: 1 });
recipeSchema.index({ 'timing.total': 1 });
recipeSchema.index({ averageRating: -1 });
recipeSchema.index({ 'stats.views': -1 });
recipeSchema.index({ createdAt: -1 });
recipeSchema.index({ creator: 1 });
recipeSchema.index({ 'nutrition.dietaryFlags': 1 });
recipeSchema.index({ 'costAnalysis.perServing': 1 });

// Virtual for total time calculation
recipeSchema.virtual('totalTime').get(function() {
  return this.timing.prep + this.timing.cook;
});

// Virtual for cost per serving
recipeSchema.virtual('costPerServing').get(function() {
  return this.costAnalysis?.perServing || 0;
});

// Pre-save middleware to calculate total time
recipeSchema.pre('save', function(next) {
  if (this.timing.prep !== undefined && this.timing.cook !== undefined) {
    this.timing.total = this.timing.prep + this.timing.cook + (this.timing.inactive || 0);
  }
  next();
});

// Method to add rating
recipeSchema.methods.addRating = function(userId, rating, review = '') {
  // Remove existing rating from this user
  this.ratings = this.ratings.filter(r => !r.user.equals(userId));
  
  // Add new rating
  this.ratings.push({
    user: userId,
    rating,
    review
  });
  
  // Recalculate average rating
  this.calculateAverageRating();
  
  return this.save();
};

// Method to calculate average rating
recipeSchema.methods.calculateAverageRating = function() {
  if (this.ratings.length === 0) {
    this.averageRating = 0;
    this.totalRatings = 0;
    return;
  }
  
  const sum = this.ratings.reduce((acc, rating) => acc + rating.rating, 0);
  this.averageRating = Math.round((sum / this.ratings.length) * 10) / 10;
  this.totalRatings = this.ratings.length;
};

// Method to increment view count
recipeSchema.methods.incrementViews = function() {
  this.stats.views += 1;
  return this.save();
};

// Method to increment save count
recipeSchema.methods.incrementSaves = function() {
  this.stats.saves += 1;
  return this.save();
};

// Method to increment share count
recipeSchema.methods.incrementShares = function() {
  this.stats.shares += 1;
  return this.save();
};

// Method to mark as cooked
recipeSchema.methods.markAsCooked = function() {
  this.stats.cooked += 1;
  return this.save();
};

// Method to calculate quality score
recipeSchema.methods.calculateQualityScore = function() {
  let score = 0;
  
  // Basic completeness (40 points)
  if (this.title) score += 5;
  if (this.description) score += 5;
  if (this.ingredients.length > 0) score += 10;
  if (this.instructions.length > 0) score += 10;
  if (this.images.length > 0) score += 5;
  if (this.nutrition) score += 5;
  
  // Detail quality (30 points)
  if (this.timing.prep && this.timing.cook) score += 10;
  if (this.equipment.length > 0) score += 5;
  if (this.tags.length > 0) score += 5;
  if (this.costAnalysis) score += 10;
  
  // User engagement (30 points)
  if (this.averageRating > 4) score += 15;
  else if (this.averageRating > 3) score += 10;
  else if (this.averageRating > 2) score += 5;
  
  if (this.stats.views > 100) score += 5;
  if (this.stats.cooked > 10) score += 5;
  if (this.verified) score += 5;
  
  this.qualityScore = Math.min(score, 100);
  return this.qualityScore;
};

// Static method to find recipes by ingredients
recipeSchema.statics.findByIngredients = function(ingredientNames, options = {}) {
  const {
    minMatch = 1,
    maxTime,
    maxCost,
    dietaryRestrictions = [],
    cuisine,
    difficulty
  } = options;
  
  const query = {
    'ingredients.name': {
      $in: ingredientNames.map(name => new RegExp(name, 'i'))
    }
  };
  
  if (maxTime) {
    query['timing.total'] = { $lte: maxTime };
  }
  
  if (maxCost) {
    query['costAnalysis.perServing'] = { $lte: maxCost };
  }
  
  if (dietaryRestrictions.length > 0) {
    query['nutrition.dietaryFlags'] = { $all: dietaryRestrictions };
  }
  
  if (cuisine) {
    query.cuisine = cuisine;
  }
  
  if (difficulty) {
    query.difficulty = difficulty;
  }
  
  return this.find(query)
    .populate('creator', 'username avatar')
    .sort({ averageRating: -1, 'stats.views': -1 });
};

// Static method for budget-friendly recipes
recipeSchema.statics.findBudgetFriendly = function(maxCostPerServing = 5) {
  return this.find({
    'costAnalysis.perServing': { $lte: maxCostPerServing },
    'costAnalysis.budgetFriendly': true
  })
    .sort({ 'costAnalysis.perServing': 1, averageRating: -1 });
};

module.exports = mongoose.model('Recipe', recipeSchema);